### PR TITLE
make cc.xml endpoint public, and only list public pipelines

### PR DESCRIPTION
**Pull request by LukeWinikates**
[add yourself]: https://github.com/concourse/governance#individual-contributors
[PR requirements]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md#pull-request-requirements

## Changes proposed by this PR

closes # 

* [x] make speculative code changes
* [x] ask for help
* [x] make tests pass
* [x] prove it only shows public pipelines

## Notes to reviewer

Hi friends - I've never made a concourse PR before, and don't know the codebase well. I am a longtime concourse user.

I want to try out the cc.xml feature, but it requires authentication. This URL should be publicly accessible, because tools like ccmenu don't implement cookie based authentication.

I don't think that this code works quite yet, but I thought there might be value putting a sketch of the solution out there already to get advice sooner.

See also #3979

## Release Note

* Public pipelines are now accessible through the [`cc.xml`](https://concourse-ci.org/observation.html#ccxml) endpoint while unauthenticated

### CODE REVIEW

**Review by taylorsilva — CHANGES_REQUESTED** [review]
Thanks for the PR!

You'll need to add a test for this in the api test suite. You'll want to add tests to this file: https://github.com/concourse/concourse/blob/d2e262bd3fa4a56be9d0d65e9df0ecebb7f1a4cf/atc/api/cc_test.go#L355-L363

You can see the current tests will need to be updated for the non-auth'd workflow.

You can run the tests by running `go test ./atc/api/` from the root of the repo, where the `go.mod` and `go.sum` files are.

When I try to run the tests against your code I get this error:

```
┌─[feat-public-ccxml][~/workspace/concourse/concourse]
└─▪ go test ./atc/api/
atc/api/ccserver/get.go:4:2: missing go.sum entry for module providing package code.cloudfoundry.org/lager (imported by github.com/concourse/concourse/atc/api/ccserver); to add:
        go get github.com/concourse/concourse/atc/api/ccserver
```

Actually not sure if this dependency should be changed or not 🤔 Should probably still be `"code.cloudfoundry.org/lager/v3"`. Not sure if your IDE made these suggestions?

I'll leave a few other comments, but adding tests is the big thing that needs to happen here.

File: atc/api/ccserver/get.go [hunk 1]
```diff
@@ -1,14 +1,14 @@
 package ccserver
 
 import (
+	"code.cloudfoundry.org/lager"
```
Comment (taylorsilva) [inline atc/api/ccserver/get.go:4]: should be pointing to v3
```suggestion
	"code.cloudfoundry.org/lager/v3"
```

**Review by taylorsilva — CHANGES_REQUESTED** [review]

**Comment by LukeWinikates** [comment]
Thanks @taylorsilva !

**Comment by LukeWinikates** [comment]
@taylorsilva how would you expect this feature to behave for teams with no public pipelines? 404? 401? 200 with empty XML?

**Comment by taylorsilva** [comment]
@LukeWinikates 200 with empty xml seems to make the most sense at first glance.

I decided to do some research to see if anyone ever did anything like this before. Came across this comment: https://github.com/concourse/concourse/pull/5308#issuecomment-600283564
Which goes over a broader idea on how to share the status of pipelines/jobs. Way out of scope in this PR, don't change what you're doing here!

The original PR that added cc.xml has some interesting tidbits about the auth stuff: https://github.com/concourse/concourse/pull/2759
Seems like other cc.xml-based products support auth and CCMenu is just one that doesn't. How interesting!

Feel free to commit whatever work in progress (WIP) stuff you have. Can always clean up the commits after.

File: atc/api/ccserver/get.go [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
+	"github.com/concourse/concourse/atc/api/accessor"
 	"net/http"
 	"net/url"
 	"time"
 
-	"code.cloudfoundry.org/lager/v3"
-
 	"github.com/concourse/concourse/atc"
 	"github.com/concourse/concourse/atc/db"
 	"github.com/tedsuo/rata"
```
Comment (taylorsilva) [inline atc/api/ccserver/get.go:14]: nit: the import are out of order. run `go fmt` on this file and any of the other files you've modified to fix this automatically for you.

File: go.mod [hunk 3]
```diff
@@ -6,6 +6,7 @@ require (
 	code.cloudfoundry.org/clock v1.1.0
 	code.cloudfoundry.org/credhub-cli v0.0.0-20230410130651-444cd52cc23d
 	code.cloudfoundry.org/garden v0.0.0-20230418162742-6c127706d54f
+	code.cloudfoundry.org/lager v2.0.0+incompatible
```
Comment (taylorsilva) [inline go.mod:9]: Can you run `go mod tidy` as well. I think this is left over from when you import `code.cloudfoundry.org/lager`

File: atc/api/cc_test.go [hunk 4]
```diff
@@ -355,10 +355,48 @@ var _ = Describe("cc.xml", func() {
...
+
+			})
+
+			It("returns 200", func() {
+				Expect(response.StatusCode).To(Equal(http.StatusOK))
 			})
 
-			It("returns 401", func() {
-				Expect(response.StatusCode).To(Equal(http.StatusUnauthorized))
+			It("lists public pipelines", func() {
```
Comment (taylorsilva) [inline atc/api/cc_test.go:391]: This looks good! Adding another test for when there's no public pipelines will be a good idea. I guess empty `<Projects></Projects>` tags would work?

You can add that test in it's own `Context(){}` block at the same level as these `It()` blocks.

**Review by taylorsilva — CHANGES_REQUESTED** [review]
looking good so far! Just add some tests, run `go fmt` and `go mod tidy` to clean some stuff up.

File: atc/api/cc_test.go (hunk 4 above)
Comment (LukeWinikates) [reply atc/api/cc_test.go]: Thanks @taylorsilva - I reorganized the blocks a little bit. Let me know what you think.

**Comment by LukeWinikates** [comment]
> I decided to do some research to see if anyone ever did anything like this before. Came across this comment: [#5308 (comment)](https://github.com/concourse/concourse/pull/5308#issuecomment-600283564) Which goes over a broader idea on how to share the status of pipelines/jobs. Way out of scope in this PR, don't change what you're doing here!

Thanks for digging! Using the public pipeline concept to determine what cc.xml shows makes intuitive sense to me. It seems as though vito was introducing an idea of uuids so that the badge/cc.xml URL wouldn't reveal the team name, but I'm not sold on the value of that.

> The original PR that added cc.xml has some interesting tidbits about the auth stuff: #2759 Seems like other cc.xml-based products support auth and CCMenu is just one that doesn't. How interesting!

I'd be interested in alternatives to CCMenu, but I haven't had luck finding one. We are having trouble managing Slack notifications as a way of observing build errors, and this was the next thing I thought to try. I couldn't find a different cc.xml client for Mac with working install instructions.

**Comment by LukeWinikates** [comment]
@taylorsilva I just force-pushed this to bring it up to date with master and remove a `go.mod` change that didn't need to be in there that was causing a merge conflict. I don't expect any more changes. 🎩

**Comment by LukeWinikates** [comment]
@taylorsilva is there anything else I can do to help get this PR ready to merge?

**Comment by taylorsilva** [comment]
@LukeWinikates sadly not, unless you can make more time in the day for me to get stuff done haha 😅 I had to give a couple interviews at work this week, so my time has been eaten up.

All I want to do is make sure the tests are behaving the way we expect them too and then this will be good to merge. Thanks for your patience!

**Review by taylorsilva — APPROVED** [review]

**Comment by taylorsilva** [comment]
Merged! Thanks again for the PR @LukeWinikates

**Comment by LukeWinikates** [comment]
thanks @taylorsilva

**Pull request merged**